File: web_version/pmugg_files/shapes3D/bspNode.js
ms.bspNode = function(stats, plane, polygons) {
	var properties = {
			bspPolygon: new ms.alternativeArray('bspPolygon', false),
			// There are three edges. 0 for the parent, 1 for the above child, 2 for the below child.
			bspEdge: new ms.requiredArray('bspEdge', false, 3),
	};
	this.node = new ms.node(this, stats, 'bspNode', properties);
	this.node.setChangeHandler('bspPolygon', this.onChanged.bind(this));
	this.node.setChangeHandler('bspEdge', this.onChanged.bind(this));
	
	var self = this;
	polygons.forEach(function(polygon) {
		self.addPolygon(polygon);
	});

	this.plane = plane;
};

ms.bspNode.prototype.getNode = function() {
	return this.node;
};

ms.bspNode.prototype.getPolygons = function() {
	return this.node.get('bspPolygon');
};

ms.bspNode.prototype.getAbove = function() {
	var edge = this.node.get('bspEdge')[1];
	return edge ? edge.getChild() : null;
};

ms.bspNode.prototype.getBelow = function() {
	var edge = this.node.get('bspEdge')[2];
	return edge ? edge.getChild() : null;
};

ms.bspNode.prototype.addPolygon = function(polygon) {
	this.node.connect(polygon);
};

ms.bspNode.prototype.onChanged = function() {
	if (this.getPolygons().length == 0 &&
		!this.getAbove() && !this.getBelow()) {
		this.node.destroy();	
	}
};

// This is only called on the root. The children are called using addNormal.
ms.bspNode.prototype.add = function(plane, polygon, isRayCast) {
	if (!this.plane) {
		this.plane = plane;
		this.addPolygon(polygon);
		return true;
	} else {
		return this.addNormal(plane, polygon, isRayCast);
	}
};

ms.bspNode.prototype.addAbove = function(plane, polygon, isRayCast) {
	var above = this.getAbove();
	if (above) {
		return above.addNormal(plane, polygon, isRayCast);
	} else {		
		if (isRayCast) {
			return [];
		}

		var stats = this.node.getStats();
		var edge = new ms.bspEdge(stats);
		var aboveNode = new ms.bspNode(stats, plane, [polygon]);
		edge.addNodes(this, aboveNode, true)
		return true;
	}
};

ms.bspNode.prototype.addBelow = function(plane, polygon, isRayCast) {
	var below = this.getBelow();
	if (below) {
		return below.addNormal(plane, polygon, isRayCast);
	} else {
		if (isRayCast) {
			return [];
		}

		var stats = this.node.getStats();
		var edge = new ms.bspEdge(stats);
		var belowNode = new ms.bspNode(stats, plane, [polygon]);
		edge.addNodes(this, belowNode, false)
		return true;
	}
};

ms.bspNode.prototype.addNormal = function(plane, polygon, isRayCast) {
	var points = polygon.getPoints();
	if (this.plane.isParallel(plane)) {
		if (this.plane.sameD(plane)) {
			if (isRayCast) {
				return [];
			}
			// TODO: Check that this works. Seems to be OK for now.
			/* if (this.getPolygons().length > 0) {
				ms.alert('Not sure if we should be adding multiple polygons to the plane.');
			} */
			this.addPolygon(polygon);
			return true;
		} else {
			if (this.plane.sign(points[0]) == 1) {
				return this.addAbove(plane, polygon, isRayCast);
			} else {
				return this.addBelow(plane, polygon, isRayCast);
			}
		}
	}
	var self = this;
	var signs = points.map(function(point) {
		return self.plane.sign(point);
	});
	var currentSign = 0;
	var firstSign = 0;
	// All the indices where it transitions between positive and negative.
	var transitions = [];
	for (var i = 0; i < signs.length; i++) {
		var s = signs[i]
		if (firstSign == 0) {
			firstSign = s;
		}
		if (currentSign == 0) {
			currentSign = s;
		} else if ((currentSign == 1 && s == -1) ||
		           (currentSign == -1 && s == 1)) {
			currentSign = s;
			transitions.push(i);
		}
	}
	if (currentSign != firstSign) {
		transitions.unshift(0);
	}
	if (currentSign == 0) {
		ms.alert('One point should be above or below the plane.');
	}
	// If there are no transitions all points are above or below the plane.
	if (transitions.length == 0) {
		if (currentSign == 1) {
			return this.addAbove(plane, polygon, isRayCast);
		} else {
			return this.addBelow(plane, polygon, isRayCast);
		}
	}
	if (transitions.length % 2 == 1) {
		ms.alert('There should be an even number of transitions.');
	}
	var N = points.length;
	var transitionPoints = [];
	for (var i = 0; i < transitions.length; i++) {
		var index = transitions[i];
		var prevIndex = (index + N - 1) % N;
		if (sign[index] == 0) {
			transitionPoints.push(points[index])
		} else if (sign[prevIndex] == 0) {
			transitionPoints.push(points[prevIndex])
		} else {
			transitionPoints.push(this.plane.crossingPoint(points[index], points[prevIndex]));
		}
	}
	
	// Sort the transition points along the line where the two planes
	// intersect. v is the direction of the line.
	var nA = plane.getN();
	var nB = this.plane.getN();
	var v = nA.cross(nB);
	var sortedPoints = transitionPoints.slice();
	sortedPoints.sort(function(a, b) {
		return v.dot(a) - v.dot(b);
	});
	var intersection = this.intersectsPolygon(sortedPoints, v, isRayCast);
	var intersections = [];
	if (intersection) {
		if (isRayCast) {
			intersections.push({position: sortedPoints[0], polygon: intersection});
		} else {
			return false;
		}
	}

	currentSign *= -1;
	var abovePoints = [];
	var belowPoints = [];
	for (var i = 0; i < transitions.length; i++) {
		var i1 = (i + 1) % transitions.length;
		var index0 = transitions[i];
		var index1 = transitions[i1];
		if (index1 < index0) {
			index1 += N;
		}
		var pointsToAdd = (currentSign == 1) ? abovePoints : belowPoints;
		pointsToAdd.push(transitionPoints[i]);
		for (var j = index0; j < index1; j++) {
			pointsToAdd.push(points[j % N]);
		}
		pointsToAdd.push(transitionPoints[i1]);
		currentSign *= -1;
	}
	if (isRayCast) {
		intersections = intersections.concat(this.addAbove(plane, polygon, isRayCast));
		intersections = intersections.concat(this.addBelow(plane, polygon, isRayCast));
		return intersections;
	} else {		
		var abovePolygon = new ms.bspPolygon(abovePoints, polygon.getFace());
		var belowPolygon = new ms.bspPolygon(belowPoints, polygon.getFace());
		var success = this.addAbove(plane, abovePolygon, isRayCast);
		return success && this.addBelow(plane, belowPolygon, isRayCast);
	}
};

ms.bspNode.prototype.intersectsPolygon = function(transitionPoints, v, isRayCast) {
	if (!isRayCast) {
		// Check if any of edges formed by the transitionPoints intersects any polygon edges.
		for (var i = 0; i < transitionPoints.length; i += 2) {
			var a0 = transitionPoints[i];
			var a1 = transitionPoints[i + 1];
			var intersects = this.getPolygons().some(function(polygon) {
				var points = polygon.getPoints();
				var n = points.length;
				for (var j = 0; j < n; j++) {
					var b0 = points[j];
					var b1 = points[(j + 1) % n];
					if (ms.intersector.intersect3D(a0, a1, b0, b1)) {
						return true;
					}
				}
			});
			if (intersects) {
				return true;
			}
		}
	}
	
	// Check if a point is completely inside the polygon. Cast a ray. If it intersects
	// the polygon an odd number of times, it is inside the polygon.
	var a0 = transitionPoints[0];
	var a1 = transitionPoints[0].copy().add(v.copy().scale(1e4));
	for (var i = 0; i < this.getPolygons().length; i++) {
		var polygon = this.getPolygons()[i];
		var points = polygon.getPoints();
		var n = points.length;
		var intersectionCount = 0
		for (var j = 0; j < n; j++) {
			var b0 = points[j];
			var b1 = points[(j + 1) % n];
			if (ms.intersector.intersect3D(a0, a1, b0, b1)) {
				intersectionCount++;
			}
		}
		var isOdd = (intersectionCount % 2 == 1);
		if (isOdd) {
			if (isRayCast) {
				return polygon;
			} else {
				return isOdd;
			}
		}
	};
};

File: cpp_version/pmugg/shapes3D/bsp_node.cpp
#include "bsp_node.h"
#include "bsp_polygon.h"
#include "bsp_edge.h"
#include <algorithm>

namespace ms {

BspNode::BspNode(const Plane& p, const std::vector<BspPolygon*>& polys)
    : plane(p) {
    edges.resize(3, nullptr);
    for (auto* poly : polys) {
        addPolygon(poly);
    }
}

BspNode* BspNode::getAbove() const {
    return getChild(1);
}

BspNode* BspNode::getBelow() const {
    return getChild(2);
}

BspNode* BspNode::getChild(int index) const {
    if (index >= edges.size() || !edges[index]) return nullptr;
    return edges[index]->getChild();
}

void BspNode::setEdge(int index, BspEdge* edge) {
    if (index >= edges.size()) edges.resize(index + 1, nullptr);
    edges[index] = edge;
}

void BspNode::addPolygon(BspPolygon* polygon) {
    polygons.push_back(polygon);
}

void BspNode::onChanged() {
    if (polygons.empty() && !getAbove() && !getBelow()) {
        delete this;
    }
}

bool BspNode::add(const Plane& newPlane, BspPolygon* polygon, bool isRayCast) {
    if (!plane.isValid()) {
        plane = newPlane;
        addPolygon(polygon);
        return true;
    }
    return addNormal(newPlane, polygon, isRayCast);
}

bool BspNode::addNormal(const Plane& newPlane, BspPolygon* polygon, bool isRayCast) {
    const auto& points = polygon->getPoints();
    
    if (plane.isParallel(newPlane)) {
        if (plane.sameD(newPlane)) {
            if (isRayCast) return false;
            addPolygon(polygon);
            return true;
        }
        if (plane.sign(points[0]) == 1) {
            return addAbove(newPlane, polygon, isRayCast);
        }
        return addBelow(newPlane, polygon, isRayCast);
    }

    std::vector<int> signs;
    for (const auto& point : points) {
        signs.push_back(plane.sign(point));
    }

    int currentSign = 0;
    int firstSign = 0;
    std::vector<int> transitions;

    for (size_t i = 0; i < signs.size(); i++) {
        int s = signs[i];
        if (firstSign == 0) firstSign = s;
        if (currentSign == 0) currentSign = s;
        else if ((currentSign == 1 && s == -1) ||
                 (currentSign == -1 && s == 1)) {
            currentSign = s;
            transitions.push_back(i);
        }
    }

    if (currentSign != firstSign) {
        transitions.insert(transitions.begin(), 0);
    }

    if (transitions.empty()) {
        if (currentSign == 1) {
            return addAbove(newPlane, polygon, isRayCast);
        }
        return addBelow(newPlane, polygon, isRayCast);
    }

    if (transitions.size() % 2 != 0) {
        throw std::runtime_error("There should be an even number of transitions");
    }

    std::vector<Vec3> transitionPoints;
    int N = points.size();

    for (int index : transitions) {
        int prevIndex = (index + N - 1) % N;
        if (signs[index] == 0) {
            transitionPoints.push_back(points[index]);
        } else if (signs[prevIndex] == 0) {
            transitionPoints.push_back(points[prevIndex]);
        } else {
            transitionPoints.push_back(
                plane.crossingPoint(points[index], points[prevIndex]));
        }
    }

    Vec3 v = newPlane.getNormal().cross(plane.getNormal());
    auto* intersection = intersectsPolygon(transitionPoints, v, isRayCast);
    
    if (intersection) {
        if (isRayCast) return false;
        return false;
    }

    std::vector<Vec3> abovePoints;
    std::vector<Vec3> belowPoints;
    currentSign *= -1;

    for (size_t i = 0; i < transitions.size(); i++) {
        size_t i1 = (i + 1) % transitions.size();
        int index0 = transitions[i];
        int index1 = transitions[i1];
        if (index1 < index0) index1 += N;

        auto& pointsToAdd = (currentSign == 1) ? abovePoints : belowPoints;
        pointsToAdd.push_back(transitionPoints[i]);
        
        for (int j = index0; j < index1; j++) {
            pointsToAdd.push_back(points[j % N]);
        }
        pointsToAdd.push_back(transitionPoints[i1]);
        
        currentSign *= -1;
    }

    if (isRayCast) {
        return addAbove(newPlane, polygon, isRayCast) ||
               addBelow(newPlane, polygon, isRayCast);
    }

    auto* abovePolygon = new BspPolygon(abovePoints, polygon->getFace());
    auto* belowPolygon = new BspPolygon(belowPoints, polygon->getFace());
    
    return addAbove(newPlane, abovePolygon, isRayCast) &&
           addBelow(newPlane, belowPolygon, isRayCast);
}

bool BspNode::addAbove(const Plane& newPlane, BspPolygon* polygon, bool isRayCast) {
    auto* above = getAbove();
    if (above) {
        return above->addNormal(newPlane, polygon, isRayCast);
    }
    
    if (isRayCast) return false;

    auto* edge = new BspEdge();
    auto* aboveNode = new BspNode(newPlane, {polygon});
    edge->addNodes(this, aboveNode, true);
    return true;
}

bool BspNode::addBelow(const Plane& newPlane, BspPolygon* polygon, bool isRayCast) {
    auto* below = getBelow();
    if (below) {
        return below->addNormal(newPlane, polygon, isRayCast);
    }
    
    if (isRayCast) return false;

    auto* edge = new BspEdge();
    auto* belowNode = new BspNode(newPlane, {polygon});
    edge->addNodes(this, belowNode, false);
    return true;
}

BspPolygon* BspNode::intersectsPolygon(const std::vector<Vec3>& transitionPoints,
                                      const Vec3& v, bool isRayCast) const {
    if (!isRayCast) {
        // Check if any edges formed by transitionPoints intersect polygon edges
        for (size_t i = 0; i < transitionPoints.size(); i += 2) {
            const Vec3& a0 = transitionPoints[i];
            const Vec3& a1 = transitionPoints[i + 1];
            
            for (const auto* polygon : polygons) {
                const auto& points = polygon->getPoints();
                for (size_t j = 0; j < points.size(); j++) {
                    const Vec3& b0 = points[j];
                    const Vec3& b1 = points[(j + 1) % points.size()];
                    if (intersect3D(a0, a1, b0, b1)) {
                        return polygon;
                    }
                }
            }
        }
    }

    // Check if a point is inside any polygon
    if (!transitionPoints.empty()) {
        Vec3 rayEnd = transitionPoints[0] + v * 1e4;
        
        for (const auto* polygon : polygons) {
            const auto& points = polygon->getPoints();
            int intersectionCount = 0;
            
            for (size_t j = 0; j < points.size(); j++) {
                const Vec3& b0 = points[j];
                const Vec3& b1 = points[(j + 1) % points.size()];
                if (intersect3D(transitionPoints[0], rayEnd, b0, b1)) {
                    intersectionCount++;
                }
            }
            
            if (intersectionCount % 2 == 1) {
                return polygon;
            }
        }
    }

    return nullptr;
}

} // namespace ms 